File: ml-benchmark/synduce/constraints/memo/count_lt.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};
Inductive TreeMemo = mleaf {Int, Int} | mnode {Int, Int, TreeMemo, TreeMemo};

memo = \t: TreeMemo.
  match t with
    mleaf {x, _} -> x
  | mnode {x, _, _, _} -> x
  end;

is_memo = fix (
  \f: TreeMemo -> Bool. \t: TreeMemo.
  match t with
    mleaf {n, x} -> and (>= n 0) (if < x 2 then == n 1 else == n 0)
  | mnode {n, a, l, r} ->
    let exp = + (if < a 2 then 1 else 0) (+ (memo l) (memo r)) in
      and (and (>= n 0) (== n exp)) (and (f l) (f r))
  end
);

repr = fix (
  \f: TreeMemo -> Tree. \t: TreeMemo.
  match t with
    mleaf {n, a} -> leaf a
  | mnode {n, a, l, r} -> node {a, f l, f r}
  end
);

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf a -> if < a 2 then 1 else 0
  | node {a, l, r} -> if < a 2 then + 1 (+ (f l) (f r)) else + (f l) (f r)
  end
);

target = fix (
  \f: TreeMemo -> Compress TreeMemo. \t: TreeMemo.
  match t with
    mleaf {n, a} -> if < a 2 then t else t
  | mnode {n, a, l, r} -> if < a 2 then t else t /*Avoid recursions*/
  end
);

/* Customized generator, not used */
gen = fix (
  \f: Tree -> TreeMemo. \t: Tree.
  match t with
    leaf a -> mleaf {if < a 2 then 1 else 0, a}
  | node {a, l, r} ->
    let res = {f l, f r} in
      mnode {+ (if < a 2 then 1 else 0) (+ (memo res.1) (memo res.2)), a, res.1, res.2}
  end
);

main = /*\t: Tree. let mt = gen t in*/
  \mt: TreeMemo.
    if is_memo mt then spec (repr (target mt)) else 0;

File: ml-benchmark/synduce/constraints/memo/count_lt.ml
type tree = Leaf of int | Node of int * tree * tree
type treememo = MLeaf of int * int | MNode of int * int * treememo * treememo

let memo t =
  match t with
  | MLeaf (x, _) -> x
  | MNode (x, _, _, _) -> x

let rec is_memo t =
  match t with
  | MLeaf (n, x) ->
      (n >= 0) && (if x < 2 then n == 1 else n == 0)
  | MNode (n, a, l, r) ->
      let exp = (if a < 2 then 1 else 0) + (memo l + memo r) in
      ((n >= 0) && (n == exp)) && (is_memo l && is_memo r)

let rec repr t =
  match t with
  | MLeaf (_, a) -> Leaf a
  | MNode (_, a, l, r) -> Node (a, repr l, repr r)

let rec spec t =
  match t with
  | Leaf a -> if a < 2 then 1 else 0
  | Node (a, l, r) ->
      if a < 2 then 1 + spec l + spec r else spec l + spec r

val target: treememo -> treememo compress
let target t =
  match t with
  | MLeaf (_, a) -> if a < 2 then t else t
  | MNode (_, a, _, _) -> if a < 2 then t else t

let rec gen t =
  match t with
  | Leaf a -> MLeaf ((if a < 2 then 1 else 0), a)
  | Node (a, l, r) ->
      let res = (gen l, gen r) in
      match res with
      | (res1, res2) ->
          MNode (((if a < 2 then 1 else 0) + (memo res1 + memo res2)), a, res1, res2)

let program mt =
  if is_memo mt then spec (repr (target mt)) else 0